### ISSUE TITLE
kinit/ktutil crashes/segmentation faults with invalid keytab

### ISSUE THREAD

**Issue by petergil** [heimdal/heimdal#277]
Tested versions: 7.1.0, 7.3.0 are affected, the latest in Ubuntu 16.04 (1.7~git20150920+dfsg-4ubuntu1) is not.

When passing a ticket file to kinit or ktutil as a keytab it is possible to make kinit/ktutil crash. For ktutil the command given affects the resulting crash.

It appears that using a long principal name will not trigger the crashes.

Steps to reproduce:

```
$ KRB5CCNAME=/tmp/foo kinit foobar
foobar@NADA.KTH.SE's Password:
$ kinit -k -t /tmp/foo
kinit: storage_set_flags called with bad vno (4)
[1]    15201 segmentation fault (core dumped)  kinit -k -t /tmp/foo
```

```
$ ktutil -k /tmp/foo list
ktutil: storage_set_flags called with bad vno (4)
/tmp/foo:

*** Error in `ktutil': munmap_chunk(): invalid pointer: 0x00007f4cd3fbe870 ***
======= Backtrace: =========
/lib64/libc.so.6(+0x712ec)[0x7f4cd3fba2ec]
/lib64/libc.so.6(+0x776a6)[0x7f4cd3fc06a6]
/usr/lib64/libasn1.so.8(free_PrincipalName+0x36)[0x7f4cd334cd26]
/usr/lib64/libasn1.so.8(free_Principal+0x9)[0x7f4cd334d2b9]
/usr/lib64/libkrb5.so.26(krb5_free_principal+0x11)[0x7f4cd4bac041]
/usr/lib64/libkrb5.so.26(krb5_kt_free_entry+0x14)[0x7f4cd4b9c264]
/usr/lib64/libkrb5.so.26(+0x4a2e3)[0x7f4cd4b9d2e3]
ktutil[0x40401b]
ktutil[0x4043a3]
ktutil[0x405406]
ktutil[0x4021fb]
/lib64/libc.so.6(__libc_start_main+0xf0)[0x7f4cd3f691e0]
ktutil[0x40229a]
======= Memory map: ========
00400000-00407000 r-xp 00000000 00:26 40994790                           /usr/bin/ktutil
00607000-00608000 r--p 00007000 00:26 40994790                           /usr/bin/ktutil
00608000-00609000 rw-p 00008000 00:26 40994790                           /usr/bin/ktutil
00bdc000-00bfd000 rw-p 00000000 00:00 0                                  [heap]
7f4cd1a92000-7f4cd1aa8000 r-xp 00000000 00:26 37982230                   /usr/lib64/gcc/x86_64-pc-linux-gnu/5.4.0/libgcc
_s.so.1
7f4cd1aa8000-7f4cd1ca7000 ---p 00016000 00:26 37982230                   /usr/lib64/gcc/x86_64-pc-linux-gnu/5.4.0/libgcc
_s.so.1
7f4cd1ca7000-7f4cd1ca8000 r--p 00015000 00:26 37982230                   /usr/lib64/gcc/x86_64-pc-linux-gnu/5.4.0/libgcc
_s.so.1
7f4cd1ca8000-7f4cd1ca9000 rw-p 00016000 00:26 37982230                   /usr/lib64/gcc/x86_64-pc-linux-gnu/5.4.0/libgcc
_s.so.1
7f4cd1ca9000-7f4cd1d00000 r-xp 00000000 00:26 35578671                   /lib64/libncurses.so.6.0
7f4cd1d00000-7f4cd1eff000 ---p 00057000 00:26 35578671                   /lib64/libncurses.so.6.0
7f4cd1eff000-7f4cd1f03000 r--p 00056000 00:26 35578671                   /lib64/libncurses.so.6.0
7f4cd1f03000-7f4cd1f04000 rw-p 0005a000 00:26 35578671                   /lib64/libncurses.so.6.0
7f4cd1f04000-7f4cd1f05000 rw-p 00000000 00:00 0
7f4cd1f05000-7f4cd1f1b000 r-xp 00000000 00:26 37258566                   /lib64/libz.so.1.2.11
7f4cd1f1b000-7f4cd211a000 ---p 00016000 00:26 37258566                   /lib64/libz.so.1.2.11
7f4cd211a000-7f4cd211b000 r--p 00015000 00:26 37258566                   /lib64/libz.so.1.2.11
7f4cd211b000-7f4cd211c000 rw-p 00016000 00:26 37258566                   /lib64/libz.so.1.2.11
7f4cd211c000-7f4cd221f000 r-xp 00000000 00:26 38206512                   /lib64/libm-2.24.so
7f4cd221f000-7f4cd241e000 ---p 00103000 00:26 38206512                   /lib64/libm-2.24.so
7f4cd241e000-7f4cd241f000 r--p 00102000 00:26 38206512                   /lib64/libm-2.24.so
7f4cd241f000-7f4cd2420000 rw-p 00103000 00:26 38206512                   /lib64/libm-2.24.so
7f4cd2420000-7f4cd2464000 r-xp 00000000 00:26 38703875                   /lib64/libreadline.so.7.0
7f4cd2464000-7f4cd2663000 ---p 00044000 00:26 38703875                   /lib64/libreadline.so.7.0
7f4cd2663000-7f4cd2665000 r--p 00043000 00:26 38703875                   /lib64/libreadline.so.7.0
7f4cd2665000-7f4cd266b000 rw-p 00045000 00:26 38703875                   /lib64/libreadline.so.7.0
7f4cd266b000-7f4cd266d000 rw-p 00000000 00:00 0
7f4cd266d000-7f4cd2680000 r-xp 00000000 00:26 38206520                   /lib64/libresolv-2.24.so
7f4cd2680000-7f4cd2880000 ---p 00013000 00:26 38206520                   /lib64/libresolv-2.24.so
7f4cd2880000-7f4cd2881000 r--p 00013000 00:26 38206520                   /lib64/libresolv-2.24.so
7f4cd2881000-7f4cd2882000 rw-p 00014000 00:26 38206520                   /lib64/libresolv-2.24.so
7f4cd2882000-7f4cd2884000 rw-p 00000000 00:00 0
7f4cd2884000-7f4cd2886000 r-xp 00000000 00:26 38206510                   /lib64/libdl-2.24.so
7f4cd2886000-7f4cd2a86000 ---p 00002000 00:26 38206510                   /lib64/libdl-2.24.so
7f4cd2a86000-7f4cd2a87000 r--p 00002000 00:26 38206510                   /lib64/libdl-2.24.so
7f4cd2a87000-7f4cd2a88000 rw-p 00003000 00:26 38206510                   /lib64/libdl-2.24.so
7f4cd2a88000-7f4cd2a90000 r-xp 00000000 00:26 38206513                   /lib64/libcrypt-2.24.so
7f4cd2a90000-7f4cd2c8f000 ---p 00008000 00:26 38206513                   /lib64/libcrypt-2.24.so
7f4cd2c8f000-7f4cd2c90000 r--p 00007000 00:26 38206513                   /lib64/libcrypt-2.24.so
7f4cd2c90000-7f4cd2c91000 rw-p 00008000 00:26 38206513                   /lib64/libcrypt-2.24.so
7f4cd2c91000-7f4cd2cbf000 rw-p 00000000 00:00 0
7f4cd2cbf000-7f4cd2ed2000 r-xp 00000000 00:26 37725028                   /usr/lib64/libcrypto.so.1.0.0
7f4cd2ed2000-7f4cd30d2000 ---p 00213000 00:26 37725028                   /usr/lib64/libcrypto.so.1.0.0
7f4cd30d2000-7f4cd30ee000 r--p 00213000 00:26 37725028                   /usr/lib64/libcrypto.so.1.0.0
7f4cd30ee000-7f4cd30fa000 rw-p 0022f000 00:26 37725028                   /usr/lib64/libcrypto.so.1.0.0
7f4cd30fa000-7f4cd30fd000 rw-p 00000000 00:00 0
7f4cd30fd000-7f4cd310b000 r-xp 00000000 00:26 40994837                   /usr/lib64/libheimbase.so.1.0.0
7f4cd310b000-7f4cd330b000 ---p 0000e000 00:26 40994837                   /usr/lib64/libheimbase.so.1.0.0
7f4cd330b000-7f4cd330c000 r--p 0000e000 00:26 40994837                   /usr/lib64/libheimbase.so.1.0.0
7f4cd330c000-7f4cd330d000 rw-p 0000f000 00:26 40994837                   /usr/lib64/libheimbase.so.1.0.0
7f4cd330d000-7f4cd33b3000 r-xp 00000000 00:26 40994843                   /usr/lib64/libasn1.so.8.0.0
7f4cd33b3000-7f4cd35b3000 ---p 000a6000 00:26 40994843                   /usr/lib64/libasn1.so.8.0.0
7f4cd35b3000-7f4cd35b4000 r--p 000a6000 00:26 40994843                   /usr/lib64/libasn1.so.8.0.0
7f4cd35b4000-7f4cd35b8000 rw-p 000a7000 00:26 40994843                   /usr/lib64/libasn1.so.8.0.0
7f4cd35b8000-7f4cd36c8000 r-xp 00000000 00:26 40522727                   /usr/lib64/libsqlite3.so.0.8.6
7f4cd36c8000-7f4cd38c8000 ---p 00110000 00:26 40522727                   /usr/lib64/libsqlite3.so.0.8.6
7f4cd38c8000-7f4cd38cb000 r--p 00110000 00:26 40522727                   /usr/lib64/libsqlite3.so.0.8.6
7f4cd38cb000-7f4cd38ce000 rw-p 00113000 00:26 40522727                   /usr/lib64/libsqlite3.so.0.8.6
7f4cd38ce000-7f4cd38f6000 r-xp 00000000 00:26 40994838                   /usr/lib64/libwind.so.0.0.0
7f4cd38f6000-7f4cd3af5000 ---p 00028000 00:26 40994838                   /usr/lib64/libwind.so.0.0.0
7f4cd3af5000-7f4cd3af6000 r--p 00027000 00:26 40994838                   /usr/lib64/libwind.so.0.0.0
7f4cd3af6000-7f4cd3af7000 rw-p 00028000 00:26 40994838                   /usr/lib64/libwind.so.0.0.0
7f4cd3af7000-7f4cd3b40000 r-xp 00000000 00:26 40994842                   /usr/lib64/libhx509.so.5.0.0
7f4cd3b40000-7f4cd3d40000 ---p 00049000 00:26 40994842                   /usr/lib64/libhx509.so.5.0.0
7f4cd3d40000-7f4cd3d42000 r--p 00049000 00:26 40994842                   /usr/lib64/libhx509.so.5.0.0
7f4cd3d42000-7f4cd3d44000 rw-p 0004b000 00:26 40994842                   /usr/lib64/libhx509.so.5.0.0
7f4cd3d44000-7f4cd3d45000 rw-p 00000000 00:00 0
7f4cd3d45000-7f4cd3d48000 r-xp 00000000 00:26 37999566                   /lib64/libcom_err.so.2.1
7f4cd3d48000-7f4cd3f47000 ---p 00003000 00:26 37999566                   /lib64/libcom_err.so.2.1
7f4cd3f47000-7f4cd3f48000 r--p 00002000 00:26 37999566                   /lib64/libcom_err.so.2.1
7f4cd3f48000-7f4cd3f49000 rw-p 00003000 00:26 37999566                   /lib64/libcom_err.so.2.1
7f4cd3f49000-7f4cd40d6000 r-xp 00000000 00:26 38206532                   /lib64/libc-2.24.so
7f4cd40d6000-7f4cd42d5000 ---p 0018d000 00:26 38206532                   /lib64/libc-2.24.so
7f4cd42d5000-7f4cd42d9000 r--p 0018c000 00:26 38206532                   /lib64/libc-2.24.so
7f4cd42d9000-7f4cd42db000 rw-p 00190000 00:26 38206532                   /lib64/libc-2.24.so
7f4cd42db000-7f4cd42df000 rw-p 00000000 00:00 0
7f4cd42df000-7f4cd42f6000 r-xp 00000000 00:26 38206514                   /lib64/libpthread-2.24.so
7f4cd42f6000-7f4cd44f5000 ---p 00017000 00:26 38206514                   /lib64/libpthread-2.24.so
7f4cd44f5000-7f4cd44f6000 r--p 00016000 00:26 38206514                   /lib64/libpthread-2.24.so
7f4cd44f6000-7f4cd44f7000 rw-p 00017000 00:26 38206514                   /lib64/libpthread-2.24.so
7f4cd44f7000-7f4cd44fb000 rw-p 00000000 00:00 0
7f4cd44fb000-7f4cd44ff000 r-xp 00000000 00:26 40994839                   /usr/lib64/libsl.so.0.2.1
7f4cd44ff000-7f4cd46ff000 ---p 00004000 00:26 40994839                   /usr/lib64/libsl.so.0.2.1
7f4cd46ff000-7f4cd4700000 r--p 00004000 00:26 40994839                   /usr/lib64/libsl.so.0.2.1
7f4cd4700000-7f4cd4701000 rw-p 00005000 00:26 40994839                   /usr/lib64/libsl.so.0.2.1
7f4cd4701000-7f4cd4716000 r-xp 00000000 00:26 40994836                   /usr/lib64/libroken.so.18.1.0
7f4cd4716000-7f4cd4916000 ---p 00015000 00:26 40994836                   /usr/lib64/libroken.so.18.1.0
7f4cd4916000-7f4cd4917000 r--p 00015000 00:26 40994836                   /usr/lib64/libroken.so.18.1.0
7f4cd4917000-7f4cd4918000 rw-p 00016000 00:26 40994836                   /usr/lib64/libroken.so.18.1.0
7f4cd4918000-7f4cd494f000 r-xp 00000000 00:26 40994841                   /usr/lib64/libhcrypto.so.4.1.0
7f4cd494f000-7f4cd4b4f000 ---p 00037000 00:26 40994841                   /usr/lib64/libhcrypto.so.4.1.0
7f4cd4b4f000-7f4cd4b50000 r--p 00037000 00:26 40994841                   /usr/lib64/libhcrypto.so.4.1.0
7f4cd4b50000-7f4cd4b52000 rw-p 00038000 00:26 40994841                   /usr/lib64/libhcrypto.so.4.1.0
7f4cd4b52000-7f4cd4b53000 rw-p 00000000 00:00 0
7f4cd4b53000-7f4cd4bdf000 r-xp 00000000 00:26 40994845                   /usr/lib64/libkrb5.so.26.0.0
7f4cd4bdf000-7f4cd4dde000 ---p 0008c000 00:26 40994845                   /usr/lib64/libkrb5.so.26.0.0
7f4cd4dde000-7f4cd4de1000 r--p 0008b000 00:26 40994845                   /usr/lib64/libkrb5.so.26.0.0
7f4cd4de1000-7f4cd4de5000 rw-p 0008e000 00:26 40994845                   /usr/lib64/libkrb5.so.26.0.0
7f4cd4de5000-7f4cd4de6000 rw-p 00000000 00:00 0
7f4cd4de6000-7f4cd4df0000 r-xp 00000000 00:26 40994846                   /usr/lib64/libkadm5clnt.so.7.0.1
7f4cd4df0000-7f4cd4fef000 ---p 0000a000 00:26 40994846                   /usr/lib64/libkadm5clnt.so.7.0.1
7f4cd4fef000-7f4cd4ff0000 r--p 00009000 00:26 40994846                   /usr/lib64/libkadm5clnt.so.7.0.1
7f4cd4ff0000-7f4cd4ff1000 rw-p 0000a000 00:26 40994846                   /usr/lib64/libkadm5clnt.so.7.0.1
7f4cd4ff1000-7f4cd5014000 r-xp 00000000 00:26 38206531                   /lib64/ld-2.24.so
7f4cd51aa000-7f4cd51b5000 rw-p 00000000 00:00 0
7f4cd5210000-7f4cd5213000 rw-p 00000000 00:00 0
7f4cd5213000-7f4cd5214000 r--p 00022000 00:26 38206531                   /lib64/ld-2.24.so
7f4cd5214000-7f4cd5215000 rw-p 00023000 00:26 38206531                   /lib64/ld-2.24.so
7f4cd5215000-7f4cd5216000 rw-p 00000000 00:00 0
7ffee1169000-7ffee118b000 rw-p 00000000 00:00 0                          [stack]
7ffee1192000-7ffee1194000 r--p 00000000 00:00 0                          [vvar]
7ffee1194000-7ffee1196000 r-xp 00000000 00:00 0                          [vdso]
ffffffffff600000-ffffffffff601000 r-xp 00000000 00:00 0                  [vsyscall]
[1]    16252 abort (core dumped)  ktutil -k /tmp/foo list
```

**Comment by elric1** [comment]
Okay, I'm having a look at this one.  I've reproduced it in some of the cases mentioned.

**Comment by elric1** [comment]
I've a fix in PR #278

**Comment by petergil** [comment]
After applying PR #278 to my local install I can no longer reproduce the errors.

Thanks for the quick fix!

**Comment by nicowilliams** [comment]
Fixed with 2e01930 by @elric1.
